### ISSUE TITLE
App crash on UWP (.Net Standard 1.3) when built in release mode and submitted to Store

### ISSUE THREAD

**Issue by MahdiGhiasi** [NLog/NLog#3436]
**NLog versions tested**: 4.6.3 and 4.5.0

_(Both versions were installed via NuGet)_

**Platform**: UWP with minimum sdk version set to 14393 ( = .Net Standard 1.3 )

**Steps to reproduce**:

I created a blank project with minimum sdk version set to 14393 and target sdk version set to 17763.

Then, I put some code in the constructor of `App` class, to initialize NLog and send a test log:

```c#
        public App()
        {
            var config = new LoggingConfiguration();

            var testTarget = new DebugTarget("testtarget")
            {
                Layout = "${logger}|${message}",
            };
            config.AddTarget(testTarget);
            config.AddRule(LogLevel.Debug, LogLevel.Fatal, testTarget);


            LogManager.Configuration = config;

            var logger = LogManager.GetCurrentClassLogger();

            logger.Warn("Test!");

            this.InitializeComponent();
            this.Suspending += OnSuspending;
        }
```

(I've also tested by adding a `FileTarget` instead of `DebugTarget`, and got exactly the same results)

When testing on debug mode, this code works fine. 

Then, I associated the app to an app in the Store and built it for uploading to the Store. The flow creates a appxbundle file for test, as well as an appxupload file for uploading to the Store.

The appxbundle test file also works fine (but Store does not accept it, saying it's a precompiled .Net Native binary and I should upload the appxupload file instead).

Then, I uploaded the appxupload to the Store and got it from the Store, it crashes on startup. I've also extracted the contents of the appxupload (there's an appxbundle there) and installed it manually, it also crashes.

I've attached Visual Studio debugger to it (from Debug->Other Debug Targets->Debug Installed App Package), and enabled exceptions. I saw these three exceptions being thrown:

```
1)

System.MethodAccessException: 'Attempt by security transparent method 'NLog.Internal.ReflectionHelpers.SafeGetTypes(System.Reflection.Assembly)' to access security critical method 'System.Reflection.AssemblyExtensions.GetTypes(System.Reflection.Assembly)' failed.

Assembly 'NLog, Version=4.0.0.0, Culture=neutral, PublicKeyToken=5120e14c03d0593c' is marked with the AllowPartiallyTrustedCallersAttribute, and uses the level 2 security transparency model.  Level 2 transparency causes all methods in AllowPartiallyTrustedCallers assemblies to become security transparent by default, which may be the cause of this exception.'

   at NLog.Internal.ReflectionHelpers.SafeGetTypes(Assembly assembly)





2)

System.ArgumentException: 'LayoutRenderer cannot be found: 'longdate''

   at NLog.Config.Factory`2.CreateInstance(String itemName)



3)

System.MethodAccessException: 'Attempt by security transparent method 'NLog.Internal.PropertyHelper.GetAllReadableProperties(System.Type)' to access security critical method 'System.Reflection.TypeExtensions.GetProperties(System.Type, System.Reflection.BindingFlags)' failed.

Assembly 'NLog, Version=4.0.0.0, Culture=neutral, PublicKeyToken=5120e14c03d0593c' is marked with the AllowPartiallyTrustedCallersAttribute, and uses the level 2 security transparency model.  Level 2 transparency causes all methods in AllowPartiallyTrustedCallers assemblies to become security transparent by default, which may be the cause of this exception.'

   at NLog.Internal.PropertyHelper.GetAllReadableProperties(Type type)
```

It seems that setting minimum sdk version to 16299 (.Net Standard 2.0) fixes the issue, but that's not always an option for apps.

(Please note that extracting appxupload and installing it manually does not work anymore when min sdk version is set to 16299, I assume something's changed in that version and Store does some processing on the package before making it available; but when submitted through the Store and downloaded from the Store, it works fine)

**Comment by 304NotModified** [comment]
You're using the .NET standard 1.3 version?

could you please also test with the xamarin build of NLog?

See the package:

![image](https://user-images.githubusercontent.com/5808377/58506120-dc20ff80-818e-11e9-8220-78e92e23c106.png)

**Comment by 304NotModified** [comment]
could you use the .NET standard 1.5 build?

**Comment by 304NotModified** [comment]
> (I've also tested by adding a `FileTarget` instead of `DebugTarget`, and got exactly the same results)

note: It cannot instantiate a target  (visible in the stacktrace) - reflection seems to be the problem.

**Comment by snakefoot** [comment]
When reading:

https://github.com/dotnet/standard/issues/312

Then it sounds like one should change this:

```c#
#if !SILVERLIGHT4
[assembly: AllowPartiallyTrustedCallers]
#endif
```

Into this:

```c#
#if !SILVERLIGHT4  && !NETSTANDARD1_0
[assembly: AllowPartiallyTrustedCallers]
#endif
```

Since `System.Reflection.TypeExtensions` for NetCore1 fails to correctly implement `AllowPartiallyTrustedCallers`. But I'm not very skilled in this area.

**Comment by 304NotModified** [comment]
Thanks for the info. Unfortunately I also know less of medium trust and related. A lot is also deprecated afaik

Nonetheless I like to fix this issue

**Comment by snakefoot** [comment]
Created #3438

**Comment by MahdiGhiasi** [comment]
> could you use the .NET standard 1.5 build of NLog?

UWP with minsdk version < 16299 only supports .Net Standard 1.4 and below, and with minsdk version >= 16299 it's .Net Standard 2.0 (and there's no problem there with NLog for .Net Standard 2.0).

Referencing NLog for .Net Standard 1.5 dll to a project win minsdk = 14393 fails (as it should), but adding it to a project with minsdk = 16299 also fails, I'm probably missing some dependencies. I'll try to fix that and see if that works.

**Comment by snakefoot** [comment]
The "funny" part is that this security-validation should only be enabled when calling NetStandard-dll from a NetFramework-dll.

The `AllowPartiallyTrustedCallers`-attribute-validation should be be skipped for NetCore-applications.

**Comment by MahdiGhiasi** [comment]
> Created #3438

I'm not familiar with the build flow of NLog at the moment, is there any documentation somewhere on how to build NLog for .Net Standard 1.3 so I can test it?

**Comment by snakefoot** [comment]
> @MahdiGhiasi is there any documentation somewhere on how to build NLog for .Net Standard 1.3 so I can test it?

When the build completes. Then there will be a pre-release-nuget-package you can try.

@MahdiGhiasi But can you verify that you are only using NetCore/NetStandard-projects in your solution? There are no projects using NetFramework?

**Comment by 304NotModified** [comment]
you could clone and switch to master (we are using gitflow). Just build the sln in src folder

**Comment by MahdiGhiasi** [comment]
> @MahdiGhiasi But can you verify that you are only using NetCore/NetStandard-projects in your solution? There are no projects using NetFramework?

This is a blank UWP project with minimum SDK verison set to 14393 and target SDK version set to 17763, and I've added the code on the first post to the constructor of App class; as far as I know it should be just NetCore.

**Comment by 304NotModified** [comment]
or do

>  dotnet build .\src\NLog\NLog.csproj -f netstandard1.3

**Comment by snakefoot** [comment]
@MahdiGhiasi Sounds very clean. There were just some comments in the linked issue, about the validation only happening for NetFramework and not for NetCore.

**Comment by 304NotModified** [comment]
> could you use the .NET standard 1.5 build of NLog?

To answer my own question ;)

Is see it's .NET standard 1.4 < 10.0.16299 and 1.5+ and 2.0 for 10.0.16299

https://docs.microsoft.com/en-us/dotnet/standard/net-standard

**Comment by 304NotModified** [comment]
> The "funny" part is that this security-validation should only be enabled when calling NetStandard-dll from a NetFramework-dll.
> 
> The `AllowPartiallyTrustedCallers`-attribute-validation should be be skipped for NetCore-applications.

We can't do that I think? :(

**Comment by snakefoot** [comment]
> We can't do that I think? :(

I'm not talking about NLog. I'm quoting the comments about how NetCore and NetFramework performs security-checking of assemblies. NLog cannot change how NetCore and NetFramework does their thing.

**Comment by MahdiGhiasi** [comment]
Sorry for the delay. I'm having some issues with adding necessary dependencies manually. Is the pre-release NuGet package created yet by the build flow? @snakefoot

**Comment by 304NotModified** [comment]
> Is the pre-release NuGet package created yet by the build flow? @snakefoot

Nope, the build queue is a bit busy as we have only 1 machine and i'm also releasing 4.6.4 ;)

**Comment by 304NotModified** [comment]
you can get the nupkg as follows:

wait for green on appveyor:

https://github.com/NLog/NLog/pull/3438

![image](https://user-images.githubusercontent.com/5808377/58514411-35465e80-81a2-11e9-965e-12d05df9fe87.png)

then press "details"

![image](https://user-images.githubusercontent.com/5808377/58514452-48f1c500-81a2-11e9-8e85-51d12d6f701b.png)

then go to artifacts tab

![image](https://user-images.githubusercontent.com/5808377/58514509-763e7300-81a2-11e9-9a8a-9a66db67ee1c.png)

**Comment by MahdiGhiasi** [comment]
Thanks.

**Comment by snakefoot** [comment]
https://ci.appveyor.com/project/nlog/nlog/builds/24877260/artifacts is now ready. See NLog.4.6.3.10087-PR3438.nupkg

**Comment by MahdiGhiasi** [comment]
The previous errors are not appearing on this build! They seemed to happen on creating `DebugTarget`, now that code succeeds, but I'm getting some missing assembly exceptions on `GetCurrentClassLogger()`.

```
Could not load file or assembly 'System.Private.Reflection.Extensibility, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
```

I thought maybe I'm doing something wrong, so I waited for the NuGet. But I'm getting the exception with that as well. However, it seems that [it is to be expected](https://stackoverflow.com/questions/50216085/missing-system-assembly-in-appxupload-only), as the `appxupload` is not completely compiled and it finalizes when you upload the package to the Store.

(Creating `FileTarget` also throws the same missing assembly exception, maybe there's a reflection method that is missing from this incomplete build that `FileTarget` uses, but `DebugTarget` doesn't?)

My *guess* is that the issue is solved now, as I'm getting a different error now and creating `DebugTarget` seem to go well, but it seems that the only way to really confirm that it works correctly is to submit the app into the Store and see how it behaves. 

This will take a while as there's a certification process; I'll submit a test to the Store and will report back as soon as I can.

**Comment by snakefoot** [comment]
@MahdiGhiasi What version of Visual Studio are you running? Could try and check for updates for Visual Studio?

**Comment by snakefoot** [comment]
@MahdiGhiasi Maybe interesting for you:

https://stackoverflow.com/questions/50216085/missing-system-assembly-in-appxupload-only

**Comment by MahdiGhiasi** [comment]
Guys, I submitted multiple submissions to the Store today, and it seems that this issue was a mistake on my part, and NLog even without the modification in #3438 works fine when submitted to the Store.

The issue was, as described in the stackoverflow page metioned in the previous posts twice, that the test case I submitted to the Store had `runFullTrust` permission enabled, and the test cases I created later was only tested locally, not submitted through Store.

There really is no proper documentation about this behavior on the related docs, so I didn't know before that `runFullTrust` causes the .Net Native to not run on the Store submission, and I observed the issue both locally and when I submitted the app to the Store, so I thought this is an issue related to the library, not incomplete compilation.

Sorry about that!
